File: DesignPatterns/01_GOF23/DP_examples/Lua/01_Creational/Prototype/ShapeCache.lua
-- 加载产品模块
require("Product")

-- 形状缓存
---@class ShapeCache
ShapeCache = {
    shapeMap = {}
}
---@param Cache ShapeCache
local function loadCache(Cache)
    Cache:RegisterPrototype(ShapeKind.Circle, Circle:new())
    Cache:RegisterPrototype(ShapeKind.Square, Square:new())
    Cache:RegisterPrototype(ShapeKind.Rectangle, Rectangle:new())
end

local instance = nil
function ShapeCache.GetInstance()
    if instance == nil then
        instance = {}
        setmetatable(instance, { __index = ShapeCache })
        loadCache(instance)
    end
    return instance
end

---@param key ShapeKind
---@param shape Shape
function ShapeCache:RegisterPrototype(key, shape)
    if (self.shapeMap[key] == nil) then
        self.shapeMap[key] = shape
    end
end

---@param key ShapeKind
function ShapeCache:RemovePrototype(key)
    self.shapeMap[key] = nil
end

---@param key ShapeKind
---@return Shape|nil
function ShapeCache:GetShape(key)
    local shape = self.shapeMap[key]
    if shape then
        return shape:clone()
    end
    return nil
end

File: DesignPatterns/01_GOF23/DP_examples/Python/01_Creational/Prototype/ShapeCache.py
from Product import ShapeKind, Circle, Square, Rectangle

class ShapeCache:
    _instance = None
    _shape_map = {}
    # def __new__(cls):
    #     if cls._instance is None:
    #         cls._instance = super().__new__(cls)
    #         cls._instance._shape_map = {}
    #     return cls._instance

    def GetInstance():
        if ShapeCache._instance is None:
            ShapeCache._instance = ShapeCache()
            ShapeCache._instance._shape_map = {}
            ShapeCache._instance.register_prototype(ShapeKind.Circle, Circle())
            ShapeCache._instance.register_prototype(ShapeKind.Square, Square())
            ShapeCache._instance.register_prototype(ShapeKind.Rectangle, Rectangle())
        return ShapeCache._instance

    def register_prototype(self, key, shape):
        self._shape_map[key] = shape

    def remove_prototype(self, key):
        if key in self._shape_map:
            del self._shape_map[key]
            return True
        return False

    def get_shape(self, key):
        if key in self._shape_map:
            return self._shape_map[key].clone()
        return None
